compositor_pipeline/src/queue/audio_queue.rs

```diff
@@ -3,99 +3,125 @@ use std::{
-        };
-        let offset = *self
-            .timestamp_offsets
-            .entry(input_id.clone())
-            .or_insert_with(|| clock_start.elapsed());
-        let start_pts = *self
-            .start_pts
-            .entry(input_id)
-            .or_insert_with(|| samples_batch.pts);
-
-        samples_batch.pts += offset;
-        samples_batch.pts -= start_pts;
-
-        input_queue.push_back(samples_batch);
-
-        Ok(())
-    }
-
-    pub fn pop_samples_set(&mut self, pts: Duration, length: Duration) -> AudioSamplesSet {
-        // Checks if any samples in batch are in [pts, pts + length] interval
-        let batch_in_range =
-            |batch: &AudioSamplesBatch| batch.pts < pts + length && batch.end() >= pts;
+    ) -> AudioSamplesSet {
+        let (start_pts, end_pts) = range;
         let samples = self
-            .input_queues
-            .iter()
-            .map(|(input_id, input_queue)| {
-                let input_samples = input_queue
-                    .iter()
-                    .filter(|batch| batch_in_range(batch))
-                    .cloned()
-                    .collect::<Vec<AudioSamplesBatch>>();
-                (input_id.clone(), input_samples)
-            })
+            .inputs
+            .iter_mut()
+            .map(|(input_id, input)| (input_id.clone(), input.pop_samples(range, clock_start)))
             .collect();
 
-        self.drop_old_samples(pts + length);
         AudioSamplesSet {
             samples,
-            pts,
-            length,
+            start_pts,
+            length: end_pts.saturating_sub(start_pts),
         }
     }
+}
+
+#[derive(Debug)]
+struct AudioQueueInput {
+    /// Frames are PTS ordered. PTS include timestamps offsets
+    queue: VecDeque<AudioSamplesBatch>,
+    /// Frames from the channel might have any PTS. When enqueuing
+    /// they need to be recalculated relative to `Queue:clock_start`.
+    receiver: Receiver<AudioSamplesBatch>,
+
+    /// Controls input initialization, buffering, and stores information
+    /// about input offset.
+    input_state: InputState<AudioSamplesBatch>,
+}
+
+impl AudioQueueInput {
+    /// Get batches that have samples in range `range` and remove them from the queue.
+    /// Batches that are partially in range will still be returned, but they won't be
+    /// removed from the queue.
+    fn pop_samples(
+        &mut self,
+        range: (Duration, Duration),
+        clock_start: Instant,
+    ) -> Vec<AudioSamplesBatch> {
+        let (start_pts, end_pts) = range;
 
-    pub fn drop_old_samples(&mut self, up_to_pts: Duration) {
-        for input_queue in self.input_queues.values_mut() {
-            while input_queue
-                .front()
-                .map_or(false, |batch| batch.end() < up_to_pts)
-            {
-                input_queue.pop_front();
+        fn is_ready(queue: &VecDeque<AudioSamplesBatch>, range_end_pts: Duration) -> bool {
```

Comment:
`is_ready` is a pretty bad name here. When I first saw function `is_ready` inside `pop_samples`, I thought that checks if enqueued samples should be pushed. Only after a while did I realize it determines if samples from channel should be read. Maybe move the whole reading from the channel part to a separate function. I wouldn't expect from function `pop_sampels` to fetch anything from the channel (even though it makes sense here), especially since the function docs don't even mention that.

Reply:
Especially since there already exists `is_ready` function in `video_queue` that does the completely different thing

Reply:
> I thought that checks if enqueued samples should be pushed

I don't follow that logic. That function checks if queue for input is ready (if it has enough samples to fill the range)